### PR TITLE
Adding initial NTLM support (from https://github.com/vision10/node-soap)

### PR DESCRIPTION
**Pull request by insanityinside**
Pulling in changes from @vision10 https://github.com/vision10/node-soap (via @r24y https://github.com/r24y/node-soap/ who did some cleanup) into one clean package for a pull request.

This is a revised version of pull #885.

Cleaned up version of #776 and #741, syntax altered to match project. I've tested it against a live NTLM SOAP server, and it seems to work perfectly. Added basic instructions to the README with respect to syntax usage with the other security types, and regenerated TOC.

There is currently a check in http.js in HttpClient.prototype.request which has an if/else for the ntlm property. I suspect there's a better way to do this, but it's how it was done in the original code. Suggestions and pointers welcome, or move code around as appropriate. The code may need refactoring, but it seems reasonable enough to work with for now, and does seem to be functioning correctly.

Also requires tests implementing.

### CODE REVIEW

**Comment by jsdevel** [comment]
@microchip is there a way to add a test?

**Comment by microchip** [comment]
I'll take a look. Should be theoretically possible.

**Comment by vision10** [comment]
I don't know how to make the test but i can provide anything needed for it

**Comment by r24y** [comment]
@microchip thanks so much for helping merge these changes in!

As far as testing, [`mockrequire`](https://www.npmjs.com/package/mockrequire) looks like it could help here. Feel free to look into it; I may have time to look at this at some point, but probably not for another week or so.

**Comment by aijanai** [comment]
guys, are there news for this?

**Comment by microchip** [comment]
@aijanai I've not had chance to do any more with this at work at the moment I'm afraid :(

**Comment by aijanai** [comment]
@microchip 

yes yes, it's perfectly ok, but can it just be merged as it is?
The problem is that there is no sane NTLM support for node anywhere. Packages lying around on NPM are just broken, incomplete or undocumented.

Since this is not breaking anything, could we please just get over the 0.7% coverage drop and release it to the public?

**Comment by microchip** [comment]
@aijanai For what it's worth, I'm using the fork I made above in production, and it seems to be working perfectly fine. Haven't tested the latest merges yet, but there's no good reason they wouldn't work.

**Comment by aijanai** [comment]
Uhm. Then there must be something I am seriously doing wrong because it keeps crashing with:
```
events.js:176
    domain.enter();
           ^

TypeError: domain.enter is not a function
    at Request.emit (events.js:176:12)
    at IncomingMessage.<anonymous> (/home/ai_ja_nai/wd-soap-proxy-lambda/node_modules/request/request.js:998:12)
    at emitOne (events.js:96:13)
    at IncomingMessage.emit (events.js:188:7)
    at IncomingMessage.Readable.read (_stream_readable.js:381:10)
    at flow (_stream_readable.js:761:34)
    at resume_ (_stream_readable.js:743:3)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)

```

My dependencies are:
```
 "dependencies": {
    "soap": "^0.18.0",
    "soap-ntlm-2": "^0.9.4"
  },
  "devDependencies": {
    "node-soap": "vision10/node-soap"
  },
```

The code is just this simple 50 lines test script:
```
var url = process.env.URL
var username = process.env.USER
var password = process.env.PASSWORD
var domain = process.env.DOMAIN

var options = {
      wsdl_options: {
                ntlm: true,
                username: username,
                password: password,
                workstation: "",
                domain: domain
            }
}

var soap = require('soap')

soap.createClient(url, options, function(err, client, body) {
  if(err){
    console.log("ERROR")
    console.log(err)
    return
  }

  client.setSecurity(new soap.NtlmSecurity(options.wsdl_options))

  console.log(client.describe())
})
```

I took it from the GitHub website and NPM page. Is this how it is supposed to look like?
Many thanks for any help

**Comment by microchip** [comment]
@aijanai Wouldn't you need to change the dependency to be microchip/node-soap to use the NTLM merged code? As I said, current code is untested, but I can't see anything that'd immediately break it. (Please note my fork is entirely unsupported, I merged in code changes someone else had made and tidied up a bit, that's it.)

**Comment by vision10** [comment]
you need to require 
```js
var soap = require('soap-ntlm-2'),
```
not the original soap

File: lib/http.js [hunk 1]
```diff
@@ -109,13 +110,31 @@ HttpClient.prototype.handleResponse = function(req, res, body) {
...
+    // sadly when using ntlm nothing to return
+    // Not sure if this can be handled in a cleaner way rather than an if/else,
+    // will try to tidy up if possible later - insanityinside
+    options.url = rurl;
+    httpNtlm[options.method.toLowerCase()](options, function (err, res) {
+      if (err) {
+        return callback(err);
+      }
+      // if result is stream
+      if( typeof res.body != 'string') {
```
Comment (herom) [inline lib/http.js]: please use `!==` for strict checl

File: lib/security/NTLMSecurity.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+"use strict";
+
+var _ = require('lodash');
+
+function NTLMSecurity(username, password, domain, workstation) {
+    if (typeof username == "object") {
```
Comment (herom) [inline lib/security/NTLMSecurity.js]: please use `===`

File: package.json [hunk 3]
```diff
@@ -14,6 +14,8 @@
     "ejs": "~2.5.5",
     "finalhandler": "^1.0.3",
     "lodash": "^3.10.1",
+    "optional": "^0.1.3",
```
Comment (herom) [inline package.json]: where do we need this dependency?

File: lib/security/NTLMSecurity.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+            username: username,
+            password: password,
+            domain: domain,
+            workstation: workstation
+        };
+    }
+}
+
+NTLMSecurity.prototype.addHeaders = function (headers) {
+  headers.Connection = 'keep-alive';
```
Comment (herom) [inline lib/security/NTLMSecurity.js]: so, is it intentional, that the method `addHeaders` does not add `headers` anywhere but sets a `Connection` property on the (hopefully available) `headers` param (Object)?

File: lib/security/NTLMSecurity.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+};
+
+NTLMSecurity.prototype.toXML = function () {
+    return '';
+};
+
+NTLMSecurity.prototype.addOptions = function (options) {
+    _.merge(options, this.defaults);
```
Comment (herom) [inline lib/security/NTLMSecurity.js]: if you call it this way, only `options` will be mutated, but `this.defaults` will stay "like it is" - is this really the behaviour you/we want?

File: package.json (hunk 3 above)
Comment (microchip) [reply package.json]: I have no idea, it came with the original modification. Happy to remove it if it's not used.

**Comment by microchip** [comment]
I'm happy to clean up the code where mentioned, but to be honest, it should probably be refactored in order to be properly testable. I currently don't have the time to do it (or the knowledge of this setup, I'm just fixing up what's already here!)

If I get chance to at some point, I'll happily do it, but in the meantime, if anyone can refactor it, I welcome pull requests.

File: lib/security/NTLMSecurity.js (hunk 4 above)
Comment (microchip) [reply lib/security/NTLMSecurity.js]: It looks that way, it's in the original code. It works the same way as in BearerSecurity, which I'd guess is what this code is based on.

Comment (herom) [reply lib/security/NTLMSecurity.js]: ah, I see.... hmmm.....

**Comment by leonfs** [comment]
Any ideas when this could get merged?

**Comment by tac-development** [comment]
@leonfs 
> Any idea when this could get merged?

If you follow the discusion in here, then it will get merged when
> Please, add tests for the new features 👍

Please feel free to contribute 🥇 😃

**Comment by herom** [comment]
@leonfs 
> Any idea when this could get merged?

We need at least a few tests for the new features in order to merge the PR (see our guidelines on [Submitting a Pull Request](https://github.com/vpulim/node-soap/blob/master/CONTRIBUTING.md#submitting-a-pull-request)). As @insanityinside said in his previous comment
> I currently don't have the time to do it (or the knowledge of this setup, I'm just fixing up what's already here!)...  I welcome pull requests.

A PR over at his fork in order to get things cleaned up and tested is very much appreciated 😺

**Comment by benthepoet** [comment]
Just opened a PR at the @insanityinside fork to fix linting issues and added a handful of tests.

**Comment by insanityinside** [comment]
@benthepoet I'll take a look at merging it this evening, then see if I can bring the codebase up to date and merge up. I've not tested it in a bit, but I've got test apps I can run with the updated code.

**Pull request merged**